### ISSUE TITLE
Feature request: context manager methods for Group

### ISSUE THREAD

**Issue by clbarnes** [zarr-developers/zarr-python#509]
A few libraries out there try to abstract over a number of different hierarchical data store groups. Some such stores (e.g. h5py's File) are best opened using a context manager (`with` statement). Some libraries (z5py, elf's knossos implementation, h5py-like/ pyn5) aim for h5py compatibility as a design goal; zarr does not, in the main, but for basic access of Groups and Arrays it can be convenient to treat them that way.

The main impediment to doing this using an abstraction layer like `elf.io` is that if you try to use a non-context-manager object like a context manager, you get attribute errors (see my workaround here: https://github.com/constantinpape/elf/blob/master/elf/io/extensions.py#L83 ).

It doesn't do any harm to throw a

```python
def __enter__(self):
    return self

def __exit__(self, *args):
    pass
```

on (could even slot a class into the inheritance hierarchy underneath MutableMapping which just adds these trivial methods, and have Stores and Groups inherit from that instead).

This doesn't actually benefit zarr-python as a library, though, and is potentially extra code noise; it's just convenient for downstream libraries trying to abstract over zarr-python as one of several backends.

**Comment by alimanfoo** [comment]
Some store classes currently do implement the context manager protocol, e.g., [ZipStore](https://github.com/zarr-developers/zarr-python/blob/58b1786fd03414d1c920cf1e19956a34bf5cfc90/zarr/storage.py#L1241), but others don't. Personally I'd be happy for all store classes to implement context manager protocol, doing this via a `Store` ABC that inherits from MutableMapping would be a nice way to go IMO.

It would probably also be fine for the `Group` class to implement the context manager protocol, calling `store.close()` on exit (which may be a no-op depending on the store).

So, e.g., add:

```python

class Store(MutableMapping):

    def close(self):
        pass

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close() 
```

...and modify `Group` to add:

```python
    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.store.close() 
```

**Comment by clbarnes** [comment]
I definitely think a Store superclass would be a good idea, as that could also implement the slow fallback behaviour for listdir and rmdir. Using core and 3rd-party Stores as interchangeably as possible would be great.

I also thought about the Group calling its store's close, but are there any potential cases where a store might want to stay open and have multiple groups dipped in and out of? Probably not, given that Groups currently don't need the context manager anyway.

**Comment by alimanfoo** [comment]
> I definitely think a Store superclass would be a good idea, as that could also implement the slow fallback behaviour for listdir and rmdir. 

Yes that would be nice.

> I also thought about the Group calling its store's close, but are there any potential cases where a store might want to stay open and have multiple groups dipped in and out of? Probably not, given that Groups currently don't need the context manager anyway.

I imagine the only case where you want to use the context manager is when you have the root group and you are handling it as you would an h5py File. In which case it should be OK to call `store.close()` I think.

**Comment by clbarnes** [comment]
The only issue with this is cases where random `MutableMapping` instances are already used as stores - for example raw `dict`s in a lot of the tests. Those obviously cannot be relied upon to have `listdir`/ `rmdir`, so the free functions still need to be used everywhere in the codebase and technically. `Group.__exit__(*args)` can still try/except `AttributeError`s arising from calling `close` on the store if it's not there.

**Comment by alimanfoo** [comment]
Ah yes, good point.

**Comment by jrbourbeau** [comment]
Closed by #510
